### PR TITLE
fix unnecessary session creation report in 2.2.x

### PR DESCRIPTION
**Pull request by andre-doherty**
Please find the pull request for issue #1085 

I am also working on pac4j 3.0.0 however it seems to be incompatible with the current bujiio master (eg HttpActionAdapter is not anymore in the http subpackage but in http.adapter subpackage which is not what Callbackfilter is waiting for

### CODE REVIEW

File: pac4j-core/src/main/java/org/pac4j/core/credentials/extractor/HeaderExtractor.java [hunk 1]
```diff
@@ -36,7 +36,7 @@ public TokenCredentials extract(WebContext context) throws HttpAction, Credentia
             throw new CredentialsException("Wrong prefix for header: " + this.headerName);
         }
 
-        final String headerWithoutPrefix = header.substring(this.prefixHeader.length());
+        final String headerWithoutPrefix = header.substring(this.prefixHeader.length()).trim();
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/credentials/extractor/HeaderExtractor.java]: As I said in another comment, let's make the trimming an option driven by a boolean property, something like `trimValue`.

File: pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java [hunk 2]
```diff
@@ -47,6 +47,12 @@ public R perform(final C context, final Config config, final HttpActionAdapter<R
         } else {
             defaultUrl = inputDefaultUrl;
         }
+        final boolean saveInSession;
+        if (inputSaveInSession == null) {
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java]: I'm a bit skeptical about the use case which justifies not to save the user profile into the web session after a remote login process: can you refresh my memory about your use case?

File: pac4j-core/src/main/java/org/pac4j/core/engine/DefaultSecurityLogic.java [hunk 3]
```diff
@@ -145,7 +149,9 @@ public R perform(final C context, final Config config, final SecurityGrantedAcce
                 } else {
                     if (startAuthentication(context, currentClients)) {
                         logger.debug("Starting authentication");
-                        saveRequestedUrl(context, currentClients);
+                        if (!ajaxRequestResolver.isAjax(context)) {
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/engine/DefaultSecurityLogic.java]: I'd prefer to have this logic `if (!ajaxRequestResolver.isAjax(context)) {` inside the `saveRequestedUrl` to make this overridable.

File: pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java (hunk 2 above)
Comment (andre-doherty) [reply pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java]: Well the use case is the following, a spa relying on a jaxws backend request authent on a saml2 idp.
Upon auth completion the idp will trigger callback filter by the mean of the browser.
The callbackfilter will then produce a samlprofile and place it in the request context. 
This profile will be changed into a jwt token and sent back to the spa client.
In this use case as the final callback is processed in one single call there is no need for a session creation. 
The spa will authenticate itself using the jwt token for further requests.

File: pac4j-core/src/main/java/org/pac4j/core/credentials/extractor/HeaderExtractor.java (hunk 1 above)
Comment (andre-doherty) [reply pac4j-core/src/main/java/org/pac4j/core/credentials/extractor/HeaderExtractor.java]: Sure, will make that change.

File: pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java (hunk 2 above)
Comment (leleuj) [reply pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java]: OK, it's a bit unusual, but it makes sense.

File: pac4j-core/src/main/java/org/pac4j/core/credentials/extractor/HeaderExtractor.java [hunk 4]
```diff
@@ -13,11 +14,49 @@
...
-    private final String headerName;
+    private String headerName;
 
-    private final String prefixHeader;
+    private String prefixHeader;
 
-    private final String clientName;
+    private String clientName;
+
+    private Boolean trimValue;
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/credentials/extractor/HeaderExtractor.java]: Just a detail, but I'd prefer to have a `boolean` here...

File: pac4j-http/src/main/java/org/pac4j/http/client/direct/HeaderClient.java [hunk 5]
```diff
@@ -52,10 +52,12 @@ public HeaderClient(final String headerName, final String prefixHeader,
 
     @Override
     protected void clientInit(final WebContext context) {
-        CommonHelper.assertNotBlank("headerName", this.headerName);
-        CommonHelper.assertNotNull("prefixHeader", this.prefixHeader);
+        if (getCredentialsExtractor() == null) {
```
Comment (leleuj) [inline pac4j-http/src/main/java/org/pac4j/http/client/direct/HeaderClient.java]: In fact, you don't need that test as the `defaultCredentialsExtractor` already does that...

**Comment by leleuj** [comment]
Two new small changes and then, we're good. You can also prepare a similar PR for the master. Thanks.

**Comment by leleuj** [comment]
I realized this change is a breaking one, so we can't accept it for the 2.2.x branch. Though, we will accept it for the master branch. Thanks

**Comment by leleuj** [comment]
Please open the same PR for the master and change this one to make it backward compatible with the 2.2.x branch (turn the `saveInProfile` input parameter into a property to keep the method signature and maybe some other adjustments). Thanks

File: pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java [hunk 6]
```diff
@@ -81,7 +93,7 @@ public R perform(final C context, final Config config, final HttpActionAdapter<R
 
             final CommonProfile profile = client.getUserProfile(credentials, context);
             logger.debug("profile: {}", profile);
-            saveUserProfile(context, config, profile, multiProfile, renewSession);
+            saveUserProfile(context, config, profile, saveInSession, multiProfile, renewSession);
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java]: Can we avoid passing the `saveInSession` property here to keep the current signature?

File: pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java [hunk 7]
```diff
@@ -93,10 +105,10 @@ public R perform(final C context, final Config config, final HttpActionAdapter<R
     }
 
     protected void saveUserProfile(final C context, final Config config, final CommonProfile profile,
-                                   final boolean multiProfile, final boolean renewSession) {
+                                   final boolean saveInSession, final boolean multiProfile, final boolean renewSession) {
         final ProfileManager manager = getProfileManager(context, config);
         if (profile != null) {
-            manager.save(true, profile, multiProfile);
+            manager.save(saveInSession, profile, multiProfile);
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java]: Same comment here: can we directly use the `saveInSession` property and no input parameter?

File: pac4j-core/src/main/java/org/pac4j/core/engine/DefaultSecurityLogic.java [hunk 8]
```diff
@@ -233,9 +237,11 @@ protected boolean startAuthentication(final C context, final List<Client> curren
      * @throws HttpAction whether an additional HTTP action is required
      */
     protected void saveRequestedUrl(final C context, final List<Client> currentClients) throws HttpAction {
-        final String requestedUrl = context.getFullRequestURL();
-        logger.debug("requestedUrl: {}", requestedUrl);
-        context.setSessionAttribute(Pac4jConstants.REQUESTED_URL, requestedUrl);
+        if (!ajaxRequestResolver.isAjax(context)) {
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/engine/DefaultSecurityLogic.java]: This is also a breaking change at runtime. Can we avoid it?

**Comment by leleuj** [comment]
@andre-doherty I added a few more comments to address.
I'm aware the process is a bit though on this PR, but I want to be sure we don't break anything as it's our commitment.

Regarding the change in Shiro, I just released buji-pac4j v3.1.0 so it should go into a new version 3.2.0.
As the right way to inject a new profile manager, you should rely on the `profileManagerFactory` property of the `Config` component and use that in the callback filter.

File: pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java (hunk 6 above)
Comment (andre-doherty) [reply pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java]: Sure, should have done it.

File: pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java (hunk 7 above)
Comment (andre-doherty) [reply pac4j-core/src/main/java/org/pac4j/core/engine/DefaultCallbackLogic.java]: OK

**Comment by andre-doherty** [comment]
Well yes, i still had one concern about the way to setup the profileManagerFactory.
As i didn't found a way to specify a lambda (ie : setProfileManagerFactory(ShiroProfileManager::new)) in shiro.ini (beanutils doens't seem to support the jdk8 and its new features), i was still wondering how to setup profileManagerFactory.

I might have to add another property (eg useShiroProfileManager) in CallbackFilter to trigger the use of shiroProfileManager in case a custom CallbackLogic implements ProfileManagerFactoryAware).

It would not break the compatibily but on the other side it would be a little too much "hardcoded" (what if one want to use a different ProfileManager).
I am still thinking about that

**Comment by leleuj** [comment]
Please account for my comments on this PR and open a specific one for the master.

Regarding Shiro, my feeling is that the `ini` format is somehow too limited and sometimes, it's better to create a new class than try to define settings via the `ini` format.

**Comment by andre-doherty** [comment]
Hello, 

Please see my latest commits. Sorry it took so long, i have been quite busy lately.
All the changes in api should be gone now, so it should be fully compatible with 2.2.x branch

The init of CallbackLogic and SecurityLogic (especially the need to set the ShiroProfileManager::new) have been addressed using two Custom classes where there only purpose is a clean setup of the above member. (Wish i had not to do it)
I havent added those to the buji lib as it was not quite clear if those belong to it, actually there is this strong dependency on buji but there natures should put those into the pac4j-core.
Please let me know if you believe they still can be added to buji. 
Note the init of the SecurityFilter and CallbackFilter would be much more cleaner than now.

Thank you

**Comment by leleuj** [comment]
I'm in line with the changes in the PR.
I'm not sure to completely follow you regarding buji-pac4j, so please submit a PR to make things clearer.

With this change on the 2.x branch, we should upgrade to version 2.3.0 (per our versions policy), so please update the version.
And submit the related PR to the master as I don't want to be behind this feature on master.

**Comment by andre-doherty** [comment]
Hi, 
I have created the 2 following pull requests : 
- one targeting the 3.0.0 pac4j branch including this change transposed into 3.0.0 branch
- one targeting the 3.2.0 buji branch (hope this pr make the above idea more explanatory).

I have also pushed some more commits on this PR (pom from 2.2.x=>2.3.0 as requested)

**Comment by leleuj** [comment]
We're good on this one. Merging. Then, I'll make some tests before cutting a release...

I have one minor comment for the PR targeting the master branch. Please have a look.

I'll review shortly the PR for Shiro.

**Comment by andre-doherty** [comment]
Hello,

Did you have the time to make the tests ? 
A release would be very much appreciated as we have apps relying on this branch going live by the end of the month.
Thank you very much

**Comment by leleuj** [comment]
I just checked the 2.3.0-SNAPSHOT branch and it seems to be working correctly. So the release is in progress... The 2.3.0 release will show up in the Maven central repository in a few hours...

**Comment by andre-doherty** [comment]
Thanks a lot!

Le ven. 16 mars 2018 à 16:39, LELEU Jérôme <notifications@github.com> a
écrit :

**Pull request merged**